### ISSUE TITLE
Suggestion: Become independent from buildandshoot.com

### ISSUE THREAD

**Issue by ordealx** [yvt/openspades#553]
Seeing that buildandshoot.com has been in constant decline over the last years and now appears to be abandoned by its owners, i want to suggest to take precautions in case it suddenly becomes completely unavailable. I think it would be great to see a independent masterserver and maybe even a proper forum to build up a community focussed on OpenSpades.

**Comment by IBPX** [comment]
On a side note, if someone could clue me in: Do we even _need_ the master server? It is very convenient to have a list, but if the master server was taken offline, could you still manually join the other servers?

I've tried to find information on the `aos://` URIs, but haven't. It seems that they go `aos://<SOME SERVER ID>:<PORT?>:<VERSION>`, with a server ID assigned by the master server. If so, that would mean you can't play without connecting to the master server, or play servers that aren't in the list.

If it is not possible to join with only the IP address? Because if not, I think that is a very important feature to be added.

On a side note, it is probably a good idea to backup the server list JSON (if I remember correctly) periodically, in case it goes down; it would likely help in building a new list.

**Comment by feikname** [comment]
@IPBX although not necessary, the master server list is a very practical feature, I'd hate to to not be able to browse the server list from OpenSpades.

Having said that, `aos://` URIs can be converted to their respective IP addressss and are independent of the master server list, there is a online tool linked in the [piqueserver](https://github.com/piqueserver/piqueserver) wiki to convert them, made by @noway421.

Also, piqueserver is a work-in-progress continuation of the current PySpades/PySnip which we started recently, so supposing people adopt it, it'd be quite possible to update the default master server list of new (and maybe the old ones) servers from there.

I think we should try to contact BuildThenSnip LLC first, to know if they plan to keep their project and the master server list, this is a kind of change that should be taken very rarely, if taken at all.

**Comment by IBPX** [comment]
@feikname Thanks! I wasn't suggesting we get rid of the master server, only take precaution so that OpenSpades is still usable without it, but as you've shown, it already is.

**Comment by VelocityRa** [comment]
@IBPX [Here](https://github.com/yvt/openspades/blob/b979647e054e09835908c2a20fba95ab1ac0c829/Sources/Core/ServerAddress.cpp#L69-L88)'s the code that handles it in OS.

There's no server ID assigned by the master server, it's just the IP encoded in a certain way. The normal way with 4 numbers seperated by dots is also acceptable as you can see in the code.

**Comment by IBPX** [comment]
> The 'normal' way with 4 numbers seperated by periods (for example 123.123.123.123) is also acceptable judging by the code.

Do you know if it's acceptable in regular AoS 0.75/0.76 as well?

**Comment by VelocityRa** [comment]
I don't know for sure, but if the code for that is there, chances are that it is.

**Comment by noway421** [comment]
aos vanilla uses integer representation of IP address in a little endian manner

**Comment by VelocityRa** [comment]
Yes that's the way pretty much all aos URIs are encoded in, but does the client also accept normal IP formatting _as well_? That's what was asked.

**Comment by IBPX** [comment]
> Yes that's the way pretty much all aos URIs that you come across are encoded in, but does the client also accept normal IP formatting as well?

Because if so, I'm not entirely sure why AoS even uses the encoded IPs in `aos://` URis, instead of normal ones.

**Comment by VelocityRa** [comment]
That's used so that clicking on a `aos://` link from a browser, or opening it from a shortcut can be made to automatically start the client. It's one great use of URIs, it's used in a lot of applications.

**Comment by IBPX** [comment]
@VelocityRa I meant why doesn't just use `aos://123.123.123.123` URIs, with normal IPs? I understand why the `aos://` URIs are used, they're very nice.

**Comment by VelocityRa** [comment]
Ah, yeah I'm not sure why really.
Two reasons of the top off my head are that it's a less compact encoding and more of a fuss to decode? The reasons we encode IPs that way is human readability only anyways, so the only reason in favour of picking that is just that; readability. People wouldn't need to read those as normal IPs anyway, (they could just share them as is) so it kinda makes sense they didn't.

**Comment by IBPX** [comment]
I suppose so.

The only reason not to do that, I guess, is that it's harder for people to take their IP and turn it into an `aos://` URI, but seeing as it seems you can use a regular "human-encoded" IP anyway, I doesn't seem to be a problem.

Thanks for the help.

**Comment by VelocityRa** [comment]
I don't think server admins would need to do that. I'm not sure about the process but most likely they just submitted their IP to the master serverlist and that was/is done automatically (or by whoever manages it).

You're welcome :)

**Comment by NotAFile** [comment]
What would be possible is for both OpenSpades and Piqueserver to get a list a list of fallback master servers from e.g. raw.github.com, so all servers switch over gracefully in case buildandshoot goes down. If somebody wants to implement something like that I wouldn't stop them but I'm not really sure it's at all necessary.

**Comment by feikname** [comment]
For the sake of information:

According to [the Internet Archive](http://web.archive.org/web/20161028232521/http://buildandshoot.com/), BuildAndShoot has entered maintenance mode since at least *October 28, 2016*. Which would mean they have been like this for **2 months and 20 days** at minimum.

**Comment by 100best** [comment]
I think a peer-to-peer master server might be nice that way no one has to host it (ex: clients host it), maybe even a peer-to-peer server would also be good because it works with closed ports (ex: routers).

**Comment by noway421** [comment]
In local network, even Bonjour/ZeroConf can be used!

Well, at least in local. Or some kind of DHT-like system. 

But probably it's not worth implementing

**Comment by noway421** [comment]
for a centralized master server some kind of BaaS/PaaS can be used: https://github.com/ripienaar/free-for-dev#paas
Some of them go for free. For example `now` offers 1GB/month bandwidth for OSS deploys.

**Comment by kmsiapps** [comment]
Mmm.. It seems like BnS website is broken now.

**Comment by feikname** [comment]
Judging by the looks seems like they're about to upload a new website.

**Comment by NotAFile** [comment]
The homepage is fine again currently, but it seems the forum has dissapeared, not sure if that is a new change

**Comment by ordealx** [comment]
the homepage is the same "under construction" type now as it was before. only showing serverlist, no forum or any other sites they used to have. it is like this for almost 4 month now with a few breaks inbetween where it was even completely unavailable for some days. if you ask me, the homepage is not fine, and it doesnt look at all like anyone is uploading a new website anywhere there. you dont need 4 month to upload a new website.

**Comment by wokawoka** [comment]
The buildandshoot master server seems down. I can't retreive the server list from the game and there are no servers listed on the public servers section of the buildandshoot.com page (it shoes just a loading servers ... screen). Does anybody know the address of an alternative master server and how to change the address in the openspades config? Thank you!

**Comment by yvt** [comment]
> Does anybody know the address of an alternative master server

None AFAIK. This [website](http://aos.acornserver.com) provides an alternate server list (probably by fetching from buildandshoot.com or having a prepopulated list) but does not provide an API endpoint. At least you can manually copy server URLs from there.

>  and how to change the address in the openspades config? 

Change `cl_serverListUrl` from `http://services.buildandshoot.com/serverlist.json` to whatever master server API endpoint you found

**Comment by wokawoka** [comment]
Alright, thank you for your answer!
Unfortunately today apparently the master servers is having problems as periodically I'm still experiencing errors in connecting to it. Let's hope that it wil lnot leave us. In that case http://aos.acornserver.com/ would be a good resource to find servers (even if I believe that most of the players would not connect without having a master server ).
Thank you again!
